src/cargo/util/flock.rs

```diff
@@ -395,8 +395,38 @@ mod sys {
 
     #[cfg(target_os = "solaris")]
     fn flock(file: &File, flag: libc::c_int) -> Result<()> {
-        // Solaris lacks flock(), so simply succeed with a no-op
-        Ok(())
+        // Solaris lacks flock(), so try to emulate using fcntl()
+        let mut flock = libc::flock {
+            l_type: 0,
+            l_whence: 0,
+            l_start: 0,
+            l_len: 0,
+            l_sysid: 0,
+            l_pid: 0,
+            l_pad: [0, 0, 0, 0],
+        };
+        if (flag & libc::LOCK_SH) != 0 {
+            flock.l_type |= libc::F_RDLCK;
+        }
+        if (flag & libc::LOCK_EX) != 0 {
+            flock.l_type |= libc::F_WRLCK;
+        }
+        if (flag & libc::LOCK_UN) != 0 {
+            flock.l_type |= libc::F_UNLCK;
+        }
```

Comment:
I don't think l_type is a bitmask.  That is, F_RDLCK|F_WRLCK would result in 3 (F_UNLCK). Should this be `=` instead? That is, something like:

```rust

flock.l_type = if flag & libc::LOCK_UN != 0 { libc::F_RDLCK }
    else if flag & libc::LOCK_EX != 0 { libc::F_WRLCK }
    else if flag & libc::LOCK_SH != 0 { libc::F_RDLCK }
    else { panic!("…") };
```

Although here those flags won't get set simultaneously, I think it looks a little strange to use bitwise-or for something that is not a bitmask.